vector/src/main/java/im/vector/app/features/home/room/detail/composer/VoiceMessageRecorderView.kt

```diff
@@ -0,0 +1,364 @@
+        }
+        lastX = currentX
+        lastY = currentY
+    }
+
+    private fun updateRecordingState(currentX: Float, currentY: Float) {
+        val distanceX = abs(firstX - currentX)
+        val distanceY = abs(firstY - currentY)
+        if (recordingState == RecordingState.STARTED) { // Determine if cancelling or locking for the first move action.
+            if (currentX < firstX && distanceX > distanceY) {
+                recordingState = RecordingState.CANCELLING
+            } else if (currentY < firstY && distanceY > distanceX) {
+                recordingState = RecordingState.LOCKING
+            }
+        } else if (recordingState == RecordingState.CANCELLING) { // Check if cancelling conditions met, also check if it should be initial state
+            if (abs(currentX - firstX) < 10 && lastX < currentX) {
+                recordingState = RecordingState.STARTED
+            } else if (shouldCancelRecording()) {
+                recordingState = RecordingState.CANCELLED
+            }
+        } else if (recordingState == RecordingState.LOCKING) { // Check if locking conditions met, also check if it should be initial state
+            if (abs(currentY - firstY) < 10 && lastY < currentY) {
+                recordingState = RecordingState.STARTED
+            } else if (shouldLockRecording()) {
+                recordingState = RecordingState.LOCKED
+            }
+        }
+    }
+
+    private fun shouldCancelRecording(): Boolean {
+        return abs(views.voiceMessageTimer.x + views.voiceMessageTimer.width - views.voiceMessageSlideToCancel.x) < 10
+    }
+
+    private fun shouldLockRecording(): Boolean {
+        return abs(views.voiceMessageLockImage.y + views.voiceMessageLockImage.height - views.voiceMessageLockArrow.y) < 10
+    }
+
+    private fun startRecordingTimer() {
+        recordingTimerTask = object : TimerTask() {
+            override fun run() {
+                recordingTime++
+                showRecordingTimer()
+                showRecordingWaveform()
+                val timeDiffToRecordingLimit = BuildConfig.VOICE_MESSAGE_DURATION_LIMIT_MS - recordingTime * 1000
+                if (timeDiffToRecordingLimit <= 0) {
+                    views.voiceMessageRecordingLayout.post {
+                        callback?.onVoiceRecordingEnded(false)
+                        recordingState = RecordingState.NONE
+                        stopRecordingTimer()
+                        hideRecordingViews(animationDuration = 0)
+                    }
+                } else if (timeDiffToRecordingLimit in 10000..11000) {
+                    views.voiceMessageRecordingLayout.post {
+                        views.voiceMessageSendButton.isVisible = false
+                        context.toast(context.getString(R.string.voice_message_n_seconds_warning_toast, (timeDiffToRecordingLimit/1000).toInt()))
```

Comment:
Do we really want to display that in a ugly toast?

Reply:
According to Figma, yes.

Reply:
I now use a TextView to apply the theme properly